rfc/instance.md

```diff
@@ -0,0 +1,145 @@
+
+- **onMutate** : A function which fires when DOM modification occurrs.
+- **schedulework**: A function to populate work queue which is to be processed in rFA callback. If we have this argument constructor will replace spaniel implementation of the schedulework. Few frameworks need all the work to be executed inside their own system. Then we need this argument.
+
+
+Only when user provides `engine` object with `onMutate` function we create an `Optimized Engine`, which will use scroll/resize/DOM change events to process scheduled work and read queues in the next earliest rFA cycle. Otherwise constructor will fallback to creating normal implementation of engine which will poll rFA very frequently.
+
+This API will provide a way to create one engine per instance. User can take advantage of this and create multiple spaniel instances, one with optimized engine and other with unoptimized engine, based on requirement.
+
+### Interface
+The spaniel instance interface looks something like below.
+
+```typescript
+
+interface SpanielIntersectionObserverInterface {
+  new(callback: Function, options: IntersectionObserverInit): SpanielIntersectionObserver;
+}
+
+interface SpanielObserverInterface {
+  new (callback: (entries: SpanielObserverEntry[]) => void, options: SpanielObserverInit): SpanielObserver;
+}
+
+interface WatcherInterface {
+  new(config: WatcherConfig): Watcher;
+}
+
+interface SpanielInstance {
+  public IntersectionObserver: IntersectionObserverClass;
+  private engine: Engine;
+  private scheduler: Scheduler;
+  public scheduleRead: (callback: Function) => void;
+  public scheduleWork: (callback: Function) => void;
+  public eventStore: EventStore;
+  public SpanielObserver: SpanielObserverInterface;
+  public Watcher: WatcherInterface;
+  getEventStore: () => EventStore;
+  on: (eventName: string, callback: (frame: FrameInterface, id: string) => void;
+  off: (eventName: string, callback: Function) => void;
+  trigger: (eventName: string, value?: any) => void;
+  getScheduler: () => Scheduler;
+  getEngine: () => Engine;
+}
+```
+- Each instance of spaniel will have its own `engine` and `scheduler`. These two properties are not supposed to be modified directly. So are making it private property of instance. To access the these two properties we will give two functions `getScheduler` and `getEngine`.
+- Each instance will have its own event store.
+- Each instance will have on, off, trigger functions to schedule/unschedule events in the instance.
+
+
+# How We Teach This
+
+Since we are allowing users to create new instances. User must changes exisiting direct usage of spaniel to creating new instances and then using it. 
```

Comment:
> Since we are allowing users to create new instances. User must changes exisiting direct usage of spaniel to creating new instances and then using it. 

As per engine configuration is now possible, users must create IntersectionObserver's from a specific spaniel engine: This ensures a given `IntersectionObserver` is configured correctly.